### PR TITLE
Lazy loading

### PR DESCRIPTION
**Pull request by holger-nutonomy**
This PR implements a primitive form of lazy loading of the database tables by overwriting Python's default `__getattribute__` method for the attributes `ego_pose`, `sample_data` and `sample_annotation`. It can be disabled by initializing `NuScenes` with `lazy=False`.

### CODE REVIEW

File: python-sdk/nuscenes/nuscenes.py [hunk 1]
```diff
@@ -110,38 +154,18 @@ def __make_reverse_index__(self, verbose: bool) -> None:
...
         self._token2ind = dict()
-        for table in self.table_names:
-            self._token2ind[table] = dict()
+        for table_name in self.table_names:
+            if table_name not in self.lazy_tables:
+                self._token2ind[table_name] = dict()
 
-            for ind, member in enumerate(getattr(self, table)):
-                self._token2ind[table][member['token']] = ind
+                for ind, member in enumerate(self.__getattribute__(table_name)):
```
Comment (Qiang-Xu) [inline python-sdk/nuscenes/nuscenes.py]: `getattr(self, table_name)`

File: python-sdk/nuscenes/nuscenes.py [hunk 2]
```diff
@@ -92,6 +92,50 @@ def table_root(self) -> str:
         """ Returns the folder where the tables are stored for the relevant version. """
         return osp.join(self.dataroot, self.version)
 
+    def __getattribute__(self, name):
```
Comment (Qiang-Xu) [inline python-sdk/nuscenes/nuscenes.py]: why not `__getattr__`?
we can do something like:
```python
def __getattr__(self, name):
    if name == table1_name:
        table = load_table()
        build_index(table)
        setattr(self, name, table)
        return getattr(self, name)
```

File: python-sdk/tutorial.ipynb [hunk 3]
```diff
@@ -13,11 +13,35 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 1,
```
Comment (sourabh-nutonomy) [inline python-sdk/tutorial.ipynb]: Revert all the changes in the notebook.

File: python-sdk/nuscenes/nuscenes.py (hunk 2 above)
Comment (holger-nutonomy) [reply python-sdk/nuscenes/nuscenes.py]: In general that looks like a good approach, but I think here we want to explicitly store the tables in `self.tables`, so I prefer to explicitly check `if name not in self.tables`.

**Comment by sourabh-nutonomy** [comment]
```
nusc = NuScenes(version='v0.5', dataroot='/data/exp/nuScenes-blurring-data/nuscenes-v0.5', verbose=True)
nusc.getind('sample_annotation', '2cd832644d09479389ed0785e5de85c9')
```

This gives an error. Maybe turn `getind()` to a private method? Or tackle this cleanly?

File: python-sdk/nuscenes/nuscenes.py [hunk 4]
```diff
@@ -40,113 +40,144 @@ def __init__(self,
...
+        load_time = time.time()
+        if self.verbose:
+            print("======\nLoading NuScenes tables for version {} ...".format(self.version))
+
+        # Load tables that do not use lazy loading.
+        for table_name in self.table_names:
+            if table_name not in self.lazy_table_names:
+                table_content = self._load_table(table_name)
+                self.__setattr__(table_name, table_content)
+        self.tables = dict()
```
Comment (sourabh-nutonomy) [inline python-sdk/nuscenes/nuscenes.py]: I don't like that we are creating all these instance variables outside the constructor. Also look at `self._token2ind` below.

**Comment by holger-nutonomy** [comment]
We identified a more fundamental problem:
`sample['data']` is only populated when `sample_data` is loaded.
Since `sample_data` is the largest table, this would require an overhaul of the entire devkit.
Therefore I am closing this PR.

**Pull request closed without merging**